rest/admin_api.go

```diff
@@ -1564,39 +1564,45 @@ func (h *handler) updatePrincipal(name string, isUser bool) error {
 	}
 
 	newInfo.Name = &internalName
-	replaced, err := h.db.UpdatePrincipal(h.ctx(), &newInfo, isUser, h.rq.Method != "POST")
+	replaced, princ, err := h.db.UpdatePrincipal(h.ctx(), &newInfo, isUser, h.rq.Method != "POST")
 	if err != nil {
 		return err
 	} else if replaced {
 		// update event
 		if isUser {
-			base.Audit(h.ctx(), base.AuditIDUserUpdate, base.AuditFields{
-				"username": internalName,
-				"roles":    newInfo.ExplicitRoleNames.ToArray(),
-				"channels": getAuditEventAccess(h.db, isUser, newInfo),
-				"db":       h.db.Name,
-			})
+			user := princ.(auth.User)
+			if user != nil {
+				base.Audit(h.ctx(), base.AuditIDUserUpdate, base.AuditFields{
+					"username": internalName,
+					"roles":    user.ExplicitRoles().AsSet().ToArray(),
```

Comment:
ExplicitRoles().AllKeys() can be used here to avoid going TimedSet->Slice->Set->Slice.